src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheckTest.java

```diff
@@ -0,0 +1,102 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import static com.google.common.truth.Truth.assertWithMessage;
+import static com.puppycrawl.tools.checkstyle.checks.coding.UnusedLocalVariableCheck.MSG_UNUSED_LOCAL_VARIABLE;
+
+import org.junit.jupiter.api.Test;
+
+import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+public class UnusedLocalVariableCheckTest extends AbstractModuleTestSupport {
```

Comment:
I am not seeing any input examples for do-while loops, lambda parameters, or pattern variables.
I am also not seeing any examples of various odd assignments like `<<=`, `>>=`, `-=`, `/=`, etc..

I would also like to see an example where a nested anonymous class has a variable name clash with the same named variable outside the anonymous class.

Also what about a variable name that matches the name of a package but is not the package used?
Example:
````
int checks = 0;
com.puppycrawl.tools.checkstyle.checks.coding.unusedlocalvariable.InputUnusedLocalVariable.testCallingMethods();
````

Reply:
> I am not seeing any input examples for do-while loops, lambda parameters, or pattern variables.

Added examples for do-while loops, lambda parameters are not classified as local variables by [JLS](https://docs.oracle.com/javase/specs/jls/se17/html/jls-4.html#jls-4.12.3), pattern variables are not supported yet, I am working on that issue, fix is a bit complicated, discussed [here](https://github.com/checkstyle/checkstyle/issues/10969#issuecomment-982684864).

> I am also not seeing any examples of various odd assignments like <<=, >>=, -=, /=, etc..

Added those examples.

>I would also like to see an example where a nested anonymous class has a variable name clash with the same named variable outside the anonymous class.

Thanks for this one, this was not supported as instance variables were not stored but I added the support for them too.

> Also what about a variable name that matches the name of a package but is not the package used?

Added.

Reply:
> lambda parameters are not classified as local variables by JLS

Yes, but what I am concerned about is name clashes. Yes your check shouldn't process lambda parameters, but what happens if a lambda parameter is given the same name as a local variable and the lambda parameter is used. Will your check accidentally assume the local variable is what is being used instead of the lambda? I want to make sure we have coverage for false positives.

> this was not supported as instance variables were not stored 

What about non-instance variables?
````
int var = 0; // unused
new MyClass() {
    public void myMethod() {
        int var = 1; // used but in different scope
        System.out.println(var);
    }
}
````

Reply:
> Yes, but what I am concerned about is name clashes. Yes your check shouldn't process lambda parameters, but what happens if a lambda parameter is given the same name as a local variable and the lambda parameter is used.

If there is a local variable in-scope then the lambda parameter cannot have the same name.

> What about non-instance variables?

This will work fine, wont give any FP's.